Gems/Terrain/Assets/Shaders/Terrain/ClipmapComputeHelpers.azsli

```diff
@@ -112,27 +121,76 @@ float3 UnpackNormal(float2 packedNormal)
     return float3(packedNormal.xy, z);
 }
 
+struct BilinearUvs
+{
+    uint2 m_u0v0;
+    uint2 m_u0v1;
+    uint2 m_u1v0;
+    uint2 m_u1v1;
+    float2 m_weight;
+};
+
+struct ClipmapLevel
+{
+    uint m_closestLevel;
+    uint m_nextLevel; // Set to clipmapStackSize if trilinear sampling doesn't apply.
+    float m_weight; // Blending weight between 2 clipmap layers.
+};
+
 // Calculate the most detailed clipmap level.
-uint CalculateClosestClipmapLevel(
+ClipmapLevel CalculateClosestClipmapLevel(
     float2 distanceFromViewPosition,
     float maxRenderDistance,
     float clipmapScaleBase,
     uint clipmapStackSize
 )
 {
+    ClipmapLevel clipmapLevel;
+
     float2 maxRenderSize = float2(maxRenderDistance, maxRenderDistance);
     // The top clipmap's render distance
     float2 minRenderSize = maxRenderSize / pow(clipmapScaleBase, (float)(clipmapStackSize - 1u));
 
-    float2 clampedDistance = clamp(minRenderSize, abs(distanceFromViewPosition), maxRenderSize);
+    float2 clampedDistance = clamp(minRenderSize, abs(distanceFromViewPosition) * RenderDistanceBias, maxRenderSize);
     float2 distanceRatio = maxRenderSize / clampedDistance;
     // Due to clamping, the max result of log is (clipmapStackSize - 1u), which falls into the first clipmap
-    uint clipmapLevel = clipmapStackSize - 1u - uint(floor(log(min(distanceRatio.x, distanceRatio.y))/log(clipmapScaleBase)));
+    clipmapLevel.m_closestLevel = clipmapStackSize - 1u - uint(floor(log(min(distanceRatio.x, distanceRatio.y))/log(clipmapScaleBase)));
+
+    float renderDistance = maxRenderDistance / pow(clipmapScaleBase, (float)(clipmapStackSize - 1u - clipmapLevel.m_closestLevel));
+    float blendingStart = renderDistance * BlendingStartFactor;
+    float blendingEnd = renderDistance / RenderDistanceBias;
+    float blendingValue = max(clampedDistance.x, clampedDistance.y);
+
+    if (blendingValue > blendingStart && blendingValue < blendingEnd)
+    {
+        clipmapLevel.m_weight = (blendingValue - blendingStart) / (blendingEnd - blendingStart);
+        clipmapLevel.m_nextLevel = clipmapLevel.m_closestLevel + 1;
+    }
+    else
+    {
+        clipmapLevel.m_weight = 0.0;
+        clipmapLevel.m_nextLevel = clipmapStackSize;
+    }
 
     return clipmapLevel;
 }
 
-uint2 CalculateClipmapUv(
+BilinearUvs GetBilinearUvs(uint2 u0v0)
+{
+    uint clipmapSize = (uint)TerrainSrg::m_clipmapData.m_clipmapSize;
+    uint u1 = (u0v0.x + 1) % clipmapSize;
```

Comment:
Do we need to pay for the ALU modulo operation here or can we rely on a wrapping sampler?

Reply:
I will give it a try in the next PR.